File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.cpp
#include <string>
#include <vector>
#include <queue>
#include <algorithm>
#include <iostream>

using namespace std;

int BFS(int cnt, int startNode, int cutNode, vector<vector<int>> lines, int n) {
    queue<int> q;
    vector<bool> visited(n+1);
    
    q.push(startNode);
    visited[startNode] = true;
    visited[cutNode] = true;
    
    while(!q.empty()) {
        int cur = q.front();
        q.pop();
        
        for(int i = 0; i < lines[cur].size(); i++) {
            int next = lines[cur][i];
                                                 
            if(visited[next])
                continue;
            
            cnt++;
            q.push(next);
            visited[next] = true;
        }
    }
    
    cout<<cnt<<' ';
    return cnt;
}

int solution(int n, vector<vector<int>> wires) {
    int answer = 101;
    
    vector<vector<int>> lines(n+1, vector<int>());
    
    for(auto wire: wires) {
        lines[wire[0]].push_back(wire[1]);
        lines[wire[1]].push_back(wire[0]);
    }
    
    for(auto wire: wires) {
        int cnt = BFS(1, wire[0], wire[1], lines, n);
        answer = min(answer, abs(cnt - (n - cnt)));
    }
    
    return answer;
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.js
function solution(n, wires) {
    let answer = n + 1;
    
    let lines = new Array(n + 1).fill().map(() => []);
    
    const BFS = (cnt, startNode, cutNode) => {
        let queue = [];
        let visited = new Array(n + 1).fill(false);
        
        queue.push(startNode);
        visited[startNode] = true;
        visited[cutNode] = true;
        
        while(queue.length !== 0) {
            let cur = queue.shift();
            
            for(const next of lines[cur]) {
                if(visited[next])
                    continue;
                
                cnt++;
                queue.push(next);
                visited[next] = true;
            }
        }
        
        return cnt;
    }
    
    for(const [a, b] of wires) {
        lines[a].push(b);
        lines[b].push(a);
    }
    
    for(const [a, b] of wires) {
        let cnt = BFS(1, a, b);
        answer = Math.min(answer, Math.abs(cnt - (n - cnt)));
    }
    
    return answer;
}